source/fuzz/transformation_replace_load_store_with_copy_memory.cpp

```diff
@@ -0,0 +1,139 @@
+    return false;
+  }
+
+  // Intermediate values of the OpLoad and the OpStore must match.
+  return load_instruction->result_id() ==
+         store_instruction->GetSingleWordOperand(1);
+}
+
+void TransformationReplaceLoadStoreWithCopyMemory::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // OpLoad and OpStore instructions must be defined.
+  auto load_instruction =
+      FindInstruction(message_.load_instruction_descriptor(), ir_context);
+  assert(load_instruction && load_instruction->opcode() == SpvOpLoad &&
+         "The required OpLoad instruction must be defined.");
+  auto store_instruction =
+      FindInstruction(message_.store_instruction_descriptor(), ir_context);
+  assert(store_instruction && store_instruction->opcode() == SpvOpStore &&
+         "The required OpStore instruction must be defined.");
+
+  // Intermediate values of the OpLoad and the OpStore must match.
+  assert(load_instruction->result_id() ==
+             store_instruction->GetSingleWordOperand(1) &&
+         "OpLoad and OpStore must refer to the same value.");
+
+  // Coherence check: Both operands must be pointers.
+
+  // Get types of ids used as the source of OpLoad and the target of OpStore.
+  auto source = ir_context->get_def_use_mgr()->GetDef(
+      load_instruction->GetSingleWordOperand(2));
+  auto target = ir_context->get_def_use_mgr()->GetDef(
+      store_instruction->GetSingleWordInOperand(0));
+  auto source_type_opcode =
+      ir_context->get_def_use_mgr()->GetDef(source->type_id())->opcode();
+  auto target_type_opcode =
+      ir_context->get_def_use_mgr()->GetDef(target->type_id())->opcode();
+
+  // Keep release-mode compilers happy. (No unused variables.)
+  (void)target;
+  (void)source;
+  (void)target_type_opcode;
+  (void)source_type_opcode;
+
+  assert(target_type_opcode == SpvOpTypePointer &&
+         source_type_opcode == SpvOpTypePointer &&
+         "The target of OpStore and the source of OpLoad must be of type "
+         "OpTypePointer");
+
+  // Coherence check: |source| and |target| must point to the same type.
+  uint32_t target_pointee_type = fuzzerutil::GetPointeeTypeIdFromPointerType(
+      ir_context, target->type_id());
+  uint32_t source_pointee_type = fuzzerutil::GetPointeeTypeIdFromPointerType(
+      ir_context, source->type_id());
+
+  (void)target_pointee_type;
+  (void)source_pointee_type;
+
+  assert(target_pointee_type == source_pointee_type &&
+         "The target of OpStore and the source of OpLoad must point to the "
+         "same type.");
+
+  // Coherence check: The first operand of the OpLoad must match the type to
+  // which the source of OpLoad points to.
+  assert(load_instruction->GetSingleWordOperand(0) == source_pointee_type &&
+         "First operand of the OpLoad must match the type to which the source "
+         "of OpLoad points to.");
+
+  // Insert the OpCopyMemory instruction before the OpStore instruction.
```

Comment:
Extra checks: I think we may need to check that there are no memory-related instructions between the OpLoad and OpStore, similar to the check in the fuzzer pass (but this time focused on just the specific OpLoad and OpStore). Or perhaps we can just assume that this is very unlikely. The check will be fairly simple though.

Reply:
I am not sure how to implement this. Should I put checks for opcodes in a for loop and iterate from OpLoad to OpStore? How to iterate over instructions in this way? In case of MemoryBarrier/MemoryNamedBarrier should I check for storage class of the variable as well?

Reply:
Can I assume that the matching OpLoad and OpStore are always in the same block?

Reply:
I think you can get an instruction iterator, or just use `instruction->NextNode()`. I think the way that you check opcodes (including memory barriers) should match the approach used in the fuzzer pass. So if you check the storage class in the fuzzer pass, you should do it here as well. You may want to add some static functions to the transformation: one that returns true if the given opcode is a memory barrier, one that returns true if the given opcode is some other memory instruction that we care about, one that returns true if a given storage is not affected by memory barriers. Or just one function (for opcodes) if you don't want to handle storage classes yet.

Reply:
@paulthomson  Does iterating using instruction iterator work when matching OpLoad and OpStore instructions are in different blocks? Is it just simply taking instruction->NextNode() until we reach the OpStore? In the fuzzer pass we assume that they are in the same block and use for(auto& instruction: block).

Reply:
I think that if the instructions are not in the same block then IsApplicable should return false. 

Thus, you can iterate from the load to store, but if you reach the end of the block, you will get nullptr and you should return false (from IsApplicable).

Reply:
Sorry, I should have put this comment under `IsApplicable`, not under `Apply`. No need to check this in Apply.